src/extension/infer.rs

```diff
@@ -1133,6 +1160,502 @@ mod test {
+
+        let mkpred = hugr.add_node_with_parent(
+            entry,
+            NodeType::open_extensions(make_opaque(
+                A,
+                FunctionType::new(vec![NAT], twoway(NAT)).with_extension_delta(&a),
+            )),
+        )?;
+
+        let bb0 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: bc.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                B,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&bc),
+            ),
+        )?;
+
+        let bb1 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT], type_row![NAT]],
+                extension_delta: b.clone(),
+            },
+            FunctionType::new(vec![NAT], twoway(NAT)),
+            make_opaque(
+                B,
+                FunctionType::new(vec![NAT], twoway(NAT)).with_extension_delta(&b),
+            ),
+        )?;
+
+        let bb10 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: c.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                C,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&c),
+            ),
+        )?;
+
+        let bb11 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: c.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                C,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&c),
+            ),
+        )?;
+
+        // Internal wiring for DFGs
+        hugr.connect(entry_in, 0, mkpred, 0)?;
+        hugr.connect(mkpred, 0, entry_out, 0)?;
+
+        // CFG Wiring
+        hugr.connect(entry, 0, bb0, 0)?;
+        hugr.connect(entry, 0, bb1, 0)?;
+        hugr.connect(bb1, 0, bb10, 0)?;
+        hugr.connect(bb1, 0, bb11, 0)?;
+
+        hugr.connect(bb0, 0, exit, 0)?;
+        hugr.connect(bb10, 0, exit, 0)?;
+        hugr.connect(bb11, 0, exit, 0)?;
+
+        hugr.infer_extensions()?;
+
+        Ok(())
+    }
+
+    #[test]
+    fn multi_entry() -> Result<(), Box<dyn Error>> {
```

Comment:
No ascii art here, but this looks like entry -> (if(P) bb0; else bb1;) -> bb2 -> exit, which seems like a strictly simpler/easier case than the above. I guess that's OK, ascii art is nice tho (even if just a three-line comment), and why "multi_entry" ??